src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,217 @@
+
+    if (rpc_clients_.count(actor_id) == 0) {
+      RAY_CHECK(actor_locations_.count(actor_id) > 0);
+      const auto &actor_location = actor_locations_[actor_id];
+      HandleActorAlive(actor_id, actor_location.first, actor_location.second);
+    }
+
+    // Submit request.
+    auto &client = rpc_clients_[actor_id];
+    return PushTask(*client, *request, task_id, num_returns);
+  } else {
+    // Actor is dead, treat the task as failure.
+    RAY_CHECK(iter->second == ActorTableData::DEAD);
+    TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+    return Status::IOError("actor is dead or being reconstructed");
+  }
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubscribeActorUpdates() {
+  // Register a callback to handle actor notifications.
+  auto actor_notification_callback = [this](const ActorID &actor_id,
+                                            const ActorTableData &actor_data) {
+    if (actor_data.state() == ActorTableData::ALIVE) {
+      RAY_LOG(INFO) << "received notification on actor alive, actor_id: " << actor_id
+                    << ", ip address: " << actor_data.ip_address()
+                    << ", port: " << actor_data.port();
+
+      std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+      actor_states_[actor_id] = actor_data.state();
+      actor_locations_[actor_id] =
+          std::make_pair(actor_data.ip_address(), actor_data.port());
+
+      // Check if this actor is the one that we're interested, if we already have
+      // a connection to the actor, or have pending requests for it, we should
+      // create a new connection.
+      if (rpc_clients_.count(actor_id) > 0 || pending_requests_.count(actor_id) > 0) {
+        HandleActorAlive(actor_id, actor_data.ip_address(), actor_data.port());
+      }
+    } else {
+      RAY_LOG(INFO) << "received notification on actor, state="
+                    << static_cast<int>(actor_data.state()) << ", actor_id: " << actor_id;
+      std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+      actor_states_[actor_id] = actor_data.state();
+    }
+  };
+
+  return gcs_client_.Actors().AsyncSubscribe(actor_notification_callback, nullptr);
+}
+
+void CoreWorkerDirectActorTaskSubmitter::HandleActorAlive(const ActorID &actor_id,
+                                                          std::string ip_address,
+                                                          int port) {
+  std::unique_ptr<rpc::DirectActorClient> grpc_client(
+      new rpc::DirectActorClient(ip_address, port, client_call_manager_));
+  // replace old rpc client if it exists.
+  rpc_clients_[actor_id] = std::move(grpc_client);
+
+  // Submit all pending requests.
+  auto &client = rpc_clients_[actor_id];
+  auto &requests = pending_requests_[actor_id];
+  while (!requests.empty()) {
+    const auto &request = *requests.front();
+    auto status =
+        PushTask(*client, *request.request_, request.task_id_, request.num_returns_);
```

Comment:
Check the status here?

Reply:
Currently we're using unary grpc and every request would have a reply, the actual failure case has already been handled in the callback of `client.PushTask`.